### PR TITLE
Vendorize Apache 2.0 Nvidia Downscaling Code

### PR DESCRIPTION
**Pull request by frodre**
The code vendorized for diffusion/downscaling from the [EDM repo](https://github.com/NVlabs/edm) is under a CC BY-NC license, which is not compatible with Apache 2.0 (as flagged by legal).  This PR vendorizes code from `physicsnemo` (https://github.com/NVIDIA/physicsnemo/tree/327d9928abc17983ad7aa3df94da9566c197c468), which is Apache 2.0.

Changes:
- `fme.downscaling.modules.unets/preconditioners/layers`, `fme.downscaling.samplers`, and `fme.diffusion.stepper` vendorize code from `physicsnemo`
- Added regression tests against previous code for validation
- Removed unused `DhariwalUNet` module

- [x] Tests added
- [ ] If dependencies changed, "deps only" image rebuilt and "latest_deps_only_image.txt" file updated

### CODE REVIEW

File: fme/downscaling/data/topography.py [hunk 1]
```diff
@@ -116,7 +110,10 @@ def get_normalized_topography(path: str, topography_name: str = "HGTsfc"):
 
     topography_normalized = (topography - topography.mean()) / topography.std()
 
-    return Topography(data=torch.tensor(topography_normalized.values), coords=coords)
+    return Topography(
+        data=torch.tensor(topography_normalized.values, dtype=torch.float32),
```
Comment (frodre) [inline fme/downscaling/data/topography.py:120]: I was getting a dtype mismatch in `test_inference.py`.  Potentially because normalization converts to float64 and `torch.tensor` preserves?

File: fme/downscaling/modules/preconditioners.py [hunk 2]
```diff
@@ -1,58 +1,28 @@
...
+from physicsnemo.models.diffusion.preconditioning import EDMPrecond as _EDMPrecond
 
 
-class EDMPrecond(torch.nn.Module):
-    def __init__(self,
-        model,                              # The underlying neural network model. Add by <gideond@allenai.org>
-        label_dim       = 0,                # Number of class labels, 0 = unconditional.
-        use_fp16        = False,            # Execute the underlying model at FP16 precision?
-        sigma_data      = 0.5,              # Expected standard deviation of the training data.
+class EDMPrecond(_EDMPrecond):
```
Comment (frodre) [inline fme/downscaling/modules/preconditioners.py]: Adapter to preserve our previous usage where we only provide an instantiated model object.  Seems like physicsnemo carries the model init info to then initialize the model.

File: fme/downscaling/modules/registry.py [hunk 3]
```diff
@@ -143,43 +143,6 @@ def build(
         )
 
 
-@dataclasses.dataclass
-class UNetRegressionDhariwal:
```
Comment (frodre) [inline fme/downscaling/modules/registry.py:147]: Removed unused UNet style so we don't increase the vendorized dependencies for now.

File: fme/downscaling/modules/unets.py [hunk 4]
```diff
@@ -256,262 +53,25 @@ def check_level_compatibility(
...
-            emb = emb + self.map_label(tmp)
-        emb = silu(emb)
-
-        # Encoder.
-        skips = []
-        for block in self.enc.values():
-            x = block(x, emb) if isinstance(block, UNetBlock) else block(x)
-            skips.append(x)
+    def forward(self, x, *args, **kwargs):
+        validate_shape(
```
Comment (frodre) [inline fme/downscaling/modules/unets.py:381]: Didn't have to do much here except recreate our validation which sources some more useful errors for the net.

File: fme/downscaling/samplers.py [hunk 5]
```diff
@@ -1,80 +1,164 @@
 """
-This file is vendorized from edm/generate.py which you can find here:
-https://github.com/NVlabs/edm/blob/008a4e5316c8e3bfe61a62f874bddba254295afb/generate.py
+This file is vendorized from physicsnemo/physicsnemo/utis/generative/stochastic_sampler.py which you can find here:
```
Comment (frodre) [inline fme/downscaling/samplers.py:2]: Vendorized because this was a more extensive difference on Nvidia's side from what we previously used.   I mostly removed the unused conditioning/label/patching functionality.

File: fme/downscaling/samplers.py [hunk 6]
```diff
@@ -1,80 +1,164 @@
...
-    x_next = latents.to(torch.float32) * t_steps[0]
-    latent_steps = [x_next]
-    for i, (t_cur, t_next) in enumerate(zip(t_steps[:-1], t_steps[1:])): # 0, ..., N-1
+    x_next = latents.to(torch.float64) * t_steps[0]
+    latent_steps = [x_next.to(latents.dtype)]
+    for i, (t_cur, t_next) in enumerate(zip(t_steps[:-1], t_steps[1:])):  # 0, ..., N-1
         x_cur = x_next
-
         # Increase noise temporarily.
-        gamma = min(S_churn / num_steps, np.sqrt(2) - 1) if S_min <= t_cur <= S_max else 0
```
Comment (frodre) [inline fme/downscaling/samplers.py:42]: We haven't been using `S_churn` > 0, but not sure of the effects of the removal of this minimum cap on the Nvidia side.

File: fme/downscaling/samplers.py [hunk 7]
```diff
@@ -1,80 +1,164 @@
...
-        latent_steps.append(x_next)
-
-    return x_next, latent_steps
-
-
-#----------------------------------------------------------------------------
-# Wrapper for torch.Generator that allows specifying a different random seed
-# for each sample in a minibatch.
-
-class StackedRandomGenerator:
```
Comment (frodre) [inline fme/downscaling/samplers.py:66]: Unused by us

File: fme/downscaling/samplers.py (hunk 6 above)
Comment (mcgibbon) [reply fme/downscaling/samplers.py:42]: This change likely is the cause of the regression test failing when S_churn is enabled.

File: fme/downscaling/samplers.py [hunk 8]
```diff
@@ -1,80 +1,164 @@
...
-"Elucidating the Design Space of Diffusion-Based Generative Models"."""
 
-from typing import Tuple
+from typing import Callable, Optional
 
-import numpy as np
 import torch
+from torch import Tensor
+
+from physicsnemo.utils.patching import GridPatching2D
```
Comment (mcgibbon) [inline fme/downscaling/samplers.py]: I think this import is unused.

File: fme/diffusion/stepper.py [hunk 9]
```diff
@@ -8,7 +8,9 @@
 
 import dacite
 import torch
+from physicsnemo.models.diffusion import Linear, PositionalEmbedding
```
Comment (mcgibbon) [inline fme/diffusion/stepper.py]: It looks like this module doesn't exist in the latest main of physicsnemo: https://github.com/NVIDIA/physicsnemo/tree/main/physicsnemo/models

diffusion_unets exists, but it doesn't have these symbols: https://github.com/NVIDIA/physicsnemo/blob/main/physicsnemo/models/diffusion_unets/__init__.py

Here is the version you use, for reference: https://github.com/NVIDIA/physicsnemo/tree/327d9928abc17983ad7aa3df94da9566c197c468/physicsnemo/models/diffusion

Comment (mcgibbon) [reply fme/diffusion/stepper.py]: Linear and PositionalEmbedding are both basic torch modules without further dependencies, they will be easy to vendorize. The song u-net will be more difficult.

File: fme/downscaling/modules/preconditioners.py (hunk 2 above)
Comment (mcgibbon) [reply fme/downscaling/modules/preconditioners.py]: I am pretty sure we can remove the ModelMetaData import if we want to vendorize EDMPrecond - we don't use its features.

File: requirements.txt [hunk 10]
```diff
@@ -1,5 +1,6 @@
 cartopy>=0.22.0
 dacite
+einops>=0.8.1
```
Comment (frodre) [inline requirements.txt:3]: MIT License

File: fme/downscaling/modules/preconditioners.py (hunk 2 above)
Comment (frodre) [reply fme/downscaling/modules/preconditioners.py]: Removed

**Comment by frodre** [comment]
@mcgibbon, given the issue loading old checkpoints and the heavy dependency of `physicsnemo`, I went ahead and vendorized the UNet and EDMPrecond in addition to the sampler.  Only update to deps was `einops` for some of the vendorized layers.

**Review by AnnaKwa — APPROVED** [review]
Thanks Andre and Jeremy, the updates and regression tests look good to me.
